### PR TITLE
Implement entry.filename option

### PR DESCRIPTION
**Pull request by smelukov**
**What kind of change does this PR introduce?**

feature

```js
module.exports = {
    entry: {
        a: "./a.js",
	b: { import: "./b.js", filename: "c.js" }
    },
}
```

**Did you add tests for your changes?**

yes

**Does this PR introduce a breaking change?**

no

**What needs to be documented once your changes are merged?**

about entry.filename

### CODE REVIEW

File: lib/Compilation.js [hunk 1]
```diff
@@ -1582,10 +1583,10 @@ class Compilation {
 
 		this.logger.time("create chunks");
 		this.hooks.beforeChunks.call();
-		for (const [name, { dependencies }] of this.entries) {
+		for (const [name, { dependencies, options }] of this.entries) {
 			const chunk = this.addChunk(name);
 			chunk.name = name;
```
Comment (sokra) [inline lib/Compilation.js]: Set `chunk.filenameTemplate = options.filename` when set.

File: lib/EntryOptionPlugin.js [hunk 2]
```diff
@@ -70,7 +70,10 @@ class EntryOptionPlugin {
 				if (typeof value === "string" || Array.isArray(value)) {
 					applyEntryItemPlugins(value, { name });
 				} else {
-					applyEntryDescriptionPlugins(value, { name });
+					applyEntryDescriptionPlugins(value, {
+						name,
+						filename: value.filename
```
Comment (sokra) [inline lib/EntryOptionPlugin.js]: This should be handled in `applyEntryDescriptionPlugins`, not here

File: lib/Chunk.js [hunk 3]
```diff
@@ -402,6 +402,22 @@ class Chunk {
 		return false;
 	}
 
+	/**
+	 * @returns {null|Entrypoint} entrypoint with this chunk as runtime
+	 */
+	getEntrypointWithThisRuntime() {
```
Comment (sokra) [inline lib/Chunk.js]: not needed

File: lib/RuntimePlugin.js [hunk 4]
```diff
@@ -173,11 +173,19 @@ class RuntimePlugin {
...
-								chunk.filenameTemplate ||
-								(chunk.isOnlyInitial()
-									? compilation.outputOptions.filename
-									: compilation.outputOptions.chunkFilename),
+							chunk => {
+								if (chunk.filenameTemplate) {
+									return chunk.filenameTemplate;
+								}
+								if (chunk.isOnlyInitial()) {
+									const entrypoint = chunk.getEntrypointWithThisRuntime();
```
Comment (sokra) [inline lib/RuntimePlugin.js]: not needed

File: lib/javascript/JavascriptModulesPlugin.js [hunk 5]
```diff
@@ -352,7 +352,10 @@ class JavascriptModulesPlugin {
 		} else if (chunk instanceof HotUpdateChunk) {
 			return outputOptions.hotUpdateChunkFilename;
 		} else if (chunk.hasRuntime() || chunk.isOnlyInitial()) {
-			return outputOptions.filename;
+			const entrypoint = chunk.getEntrypointWithThisRuntime();
```
Comment (sokra) [inline lib/javascript/JavascriptModulesPlugin.js]: not needed

File: test/WebpackOptionsDefaulter.unittest.js [hunk 6]
```diff
@@ -155,7 +155,7 @@ describe("WebpackOptionsDefaulter", () => {
 		          "idHint": "vendors",
 		          "priority": -10,
 		          "reuseExistingChunk": true,
-		          "test": /\\[\\\\\\\\/\\]node_modules\\[\\\\\\\\/\\]/i,
+		          "test": /\\[\\\\\\\\\\\\/\\]node_modules\\[\\\\\\\\\\\\/\\]/i,
```
Comment (vankop) [inline test/WebpackOptionsDefaulter.unittest.js]: Also got problems with this snapshot..

Comment (sokra) [reply test/WebpackOptionsDefaulter.unittest.js]: But it fails on ci now. This was correct before.

Comment (vankop) [reply test/WebpackOptionsDefaulter.unittest.js]: Looks like there is a cross platform issue..

**Comment by sokra** [comment]
Thanks

**Pull request merged**